File: Go/algorithm/src/heap_sort.go
package src

func HeapSort(values []int) (ret []int, err error) {
	heapSize := len(values)
	for heapSize > 0 {
		sortHeap(values, heapSize)
		temp := values[0]
		values[0] = values[heapSize-1]
		values[heapSize-1] = temp
		heapSize -= 1
	}
	return values, nil
}

func sortHeap(values []int, heapSize int) {
	root := heapSize / 2
	for root >= 1 {
		maxHeap(values, root, heapSize)
		root -= 1
	}
}

func maxHeap(values []int, root int, heapSize int) {
	left := root * 2
	right := root*2 + 1
	maxIndex := root
	if (left <= heapSize) && (values[left-1] > values[maxIndex-1]) {
		maxIndex = left
	}
	if (right <= heapSize) && (values[right-1] > values[maxIndex-1]) {
		maxIndex = right
	}
	if maxIndex != root {
		temp := values[maxIndex-1]
		values[maxIndex-1] = values[root-1]
		values[root-1] = temp
	}
}

File: Python/algorithm/src/heap_sort.py
def max_heap(values, root, heap_size):
    left = root * 2
    right = root * 2 + 1
    max_index = root
    if (left <= heap_size) and (values[left - 1] > values[max_index - 1]):
        max_index = left
    if (right <= heap_size) and (values[right - 1] > values[max_index - 1]):
        max_index = right

    if max_index != root:
        temp = values[root - 1]
        values[root - 1] = values[max_index - 1]
        values[max_index - 1] = temp


def sort_heap(values, heap_size):
    root = heap_size // 2
    while root >= 1:
        max_heap(values, root, heap_size)
        root -= 1


def heap_sort(values):
    heap_size = len(values)
    while heap_size > 0:
        sort_heap(values, heap_size)
        temp = values[0]
        values[0] = values[heap_size - 1]
        values[heap_size - 1] = temp
        heap_size -= 1
